services/console-proxy/server/src/com/cloud/consoleproxy/WebSocketHandlerForNovnc.java

```diff
@@ -0,0 +1,505 @@
+                break;
+            }
+
+            case AUTH_RESULT_SENT: {
+                os.write(data);
+                os.flush();
+                break;
+            }
+
+            default: {
+                os.write(data);
+                os.flush();
+                break;
+            }
+        }
+
+        if (f.getType().equals(Frame.Type.CLOSE)){
+            shutdown();
+        }
+    }
+
+    /**
+     * VNC authentication.
+     */
+    private void doAuth(int authType) throws IOException {
+
+    }
+
+    /**
+     * Encode client password and send it to server.
+     */
+    private void doVncAuth(String password) throws IOException {
+
+        // Read challenge
+        byte[] challenge = new byte[16];
+        is.readFully(challenge);
+        // Encode challenge with password
+        byte[] response;
+        try {
+            response = encodePassword(challenge, password);
+        } catch (Exception e) {
+            s_logger.error("Cannot encrypt client password to send to server: " + e.getMessage());
+            throw new RuntimeException("Cannot encrypt client password to send to server: " + e.getMessage());
+        }
+
+        // Send encoded challenge
+        os.write(response);
+        os.flush();
+
+        // Read security result
+        int authResult = is.readInt();
+        switch (authResult) {
+            case RfbConstants.VNC_AUTH_OK: {
+                // Nothing to do
+                break;
+            }
+
+            case RfbConstants.VNC_AUTH_TOO_MANY:
+                s_logger.error("Connection to VNC server failed: too many wrong attempts.");
+                throw new RuntimeException("Connection to VNC server failed: too many wrong attempts.");
+
+            case RfbConstants.VNC_AUTH_FAILED:
+                s_logger.error("Connection to VNC server failed: wrong password.");
+                throw new RuntimeException("Connection to VNC server failed: wrong password.");
+
+            default:
+                s_logger.error("Connection to VNC server failed, reason code: " + authResult);
+                throw new RuntimeException("Connection to VNC server failed, reason code: " + authResult);
+        }
+    }
+
+    /**
+     * Handshake with VNC server.
+     */
+    private void handshake(DataInputStream is, DataOutputStream os) throws IOException {
+
+        // Read protocol version
+        byte[] buf = new byte[12];
+        is.readFully(buf);
+        String rfbProtocol = new String(buf);
+
+        // Server should use RFB protocol 3.x
+        if (!rfbProtocol.contains(RfbConstants.RFB_PROTOCOL_VERSION_MAJOR)) {
+            s_logger.error("Cannot handshake with VNC server. Unsupported protocol version: \"" + rfbProtocol + "\".");
+            throw new RuntimeException("Cannot handshake with VNC server. Unsupported protocol version: \"" + rfbProtocol + "\".");
+        }
+
+        sendResponseBytes(session, buf, 12);
```

Comment:
seperate client and server handshake. See the file that I attached in earlier comments